proxy/src/telemetry/sensor/http.rs

```diff
@@ -574,3 +595,48 @@ impl BodySensor for RequestBodyInner {
         &mut self.bytes_sent
     }
 }
+
+impl<S> TimestampRequestOpen<S> {
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S, B> Service for TimestampRequestOpen<S>
+where
+    S: Service<Request = http::Request<B>>,
+{
+    type Request = http::Request<B>;
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut req: Self::Request) -> Self::Future {
+        let request_open = Instant::now();
+        req.extensions_mut().insert(RequestOpen(request_open));
```

Comment:
If the intent is that we really capture the earliest instant that we have the request, then why not just add this extension at the time we constructed `req`, instead of here? In particular, it's really unclear to me what happens between the time we constructed `req` and the time `call()` is called here, and it's unclear how we know that there are no non-trivial delays between those two things.

In other words, while it seems clean to implement this as middleware, I'm not convinced that implementing it as middleware most accurately captures the data.

Reply:
@briansmith I agree with your concerns; however, it appears that the request is constructed in a library (either in `h2` [here](https://github.com/carllerche/h2/blob/fabae35311362cf24fa9e2e08e27c3d64f6c67b1/src/server.rs#L493) or in `hyper` [here](https://github.com/hyperium/hyper/blob/dfdca25c004aeb583b931f57e19510b3096362e3/src/proto/h1/dispatch.rs#L355-L359), depending on the request protocol). The middleware is the earliest point that _Conduit_ has access to the request struct.

I didn't think that PRs against those libraries in order to add a Conduit-specific extension seemed likely to be accepted, and therefore, forking our library dependencies didn't seem worth what appears to be a fairly small gain in accuracy, as the service is called with the constructed request shortly thereafter. I _did_ consider using the [`serve_modified`](https://tower-rs.github.io/tower-h2/tower_h2/server/struct.Server.html#method.serve_modified) method in `tower-h2` to add the timestamp header to the request earlier in the pipeline, but there's no similar method in `hyper`, so there would be a difference in accuracy between the HTTP/1 and HTTP/2 timestamps. I felt that using a middleware, while less accurate overall, results in more consistent measurements between both protocols.

If you have any ideas I may have overlooked, I'd appreciate your suggestion, and I'd also like to hear what @olix0r thinks, as well.

Reply:
OK. Thinking in terms of middleware, then, wouldn't the ideal be for the layer that adds the request extension to be the later directly wrapping the hyper/h2 layers?

Reply:
I'm not sure if I entirely understand what you mean --- by placing this middleware at the top of the stack (modulo https://github.com/runconduit/conduit/pull/744#discussion_r180920549), it should be among the first layers to see the request object once its' been constructed.

Reply:
> I'm not sure if I entirely understand what you mean --- by placing this middleware at the top of the stack (modulo #744 (comment)), it should be among the first layers to see the request object once its' been constructed.

How could we ensure that it is definitely the first layer to see the request object?

Reply:
1. We wrap the rest of the `NewService` stack with the `TimestampRequestOpen` `NewService`, so it's at the top of the stack:
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/lib.rs#L374
2. that stack of `NewService`s is passed to `Server::new`, and goes in the constructed `Server`'s `new_service` field 
3. when the `Server` serves an HTTP/1 request, it calls `new_service` on that stack, constructing a new `Service` where the `TimestampRequestOpen` middleware is the outermost service (the first to see the request object). It's wrapped in an adapter that turns a `tower::Service` into a `hyper::Service`, but we can't put it before that because  the version of `hyper` we depend on uses the `hyper::Request` type rather than `http::Request`, and we can't add extensions to a `hyper::Request`:
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/transparency/server.rs#L155-L169
4. When the server is constructed, it constructs a `tower_h2::Server` with the `stack`
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/transparency/server.rs#L73-L81
which is used to serve HTTP/2 connections:
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/transparency/server.rs#L178

---

In both cases, there are protocol-specific adapter middlewares around the rest of the middleware stack (`HttpBodyService` and `HyperServerService`),. In the case of the `HyperServerService` for HTTP/1, we can't add the timestamp before that service, for the reason discussed above. The H2  `HttpBodyService` just does a type conversion on the request body. We could place the middleware before that service, although I don't think it has a significant impact on latency, and this would complicate the implementation. What do you think?

Reply:
> How could we ensure that it is definitely the first layer to see the request object?

I don't think it should be the _first_ layer. (I think the error-handling/logging layer should be the first layer, so it can capture any errors emitted by middlewares like this); though it's true that the timestamping middleware can't fail on its own, so this is all semantics.

For now, I think the best we can do to _ensure_ that this is early in the stack is to document it well.

In the future, I hope that we can have better stacking primitives so these assurances can be made more explicitly (as in finagle stacks).

Reply:
i meant "error logging." I don't feel super strongly about this -- it makes some sense to just put this at the top of the stack. I don't think we have to spend too much time trying to build solid assurances that this is always at the top (yet) -- though I agree that we should at least begin to document the stack ordering.